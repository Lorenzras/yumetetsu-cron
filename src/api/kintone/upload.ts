import {kintoneClient} from '.';


export const uploadAll = (appId: string, records : Records) => {
  return kintoneClient.record.addAllRecords({
    app: appId,
    records,
  });
};

