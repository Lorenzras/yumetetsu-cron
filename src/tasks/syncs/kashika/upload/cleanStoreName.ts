export const cleanStoreName = (name: string) => {
  return name.replaceAll(/(ハウスドゥ[!！]?(\s+)?)|店/g, '');
};
