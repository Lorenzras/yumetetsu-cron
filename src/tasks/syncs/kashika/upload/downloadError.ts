import {Page} from 'puppeteer';
import fs from 'fs';
import path from 'path';
import {KStoreSettings, TStoreSettingsItem} from '../../../../config';
import {logger} from '../../../../utils';
import iconv from 'iconv-lite';

export const downloadError = async (
  page: Page,
  storeId: KStoreSettings,
) => {
  const result = await page.evaluate(()=>{
    const requestURL = $('a[href*="error"]:first').prop('href');
    return fetch(requestURL, {
      method: 'GET',
      credentials: 'include',
      headers: [['charset', 'ISO-8859-1']],
    })
      .then((res) => res.arrayBuffer())
      .then((buffer)=>{
        const decoder = new TextDecoder('shift-jis');

        const text = decoder
          .decode(buffer);

        return text;
      });
  });

  const downloadDir = path.join(__dirname, 'resultDir');

  fs.existsSync(downloadDir) || fs.mkdirSync(downloadDir, {recursive: true});

  const filePath = path
    .join(
      downloadDir,
      `${storeId}-error.csv`,
    );

  logger.info(`Saving to ${filePath}`);

  fs.writeFileSync(
    filePath,
    iconv.encode(result, 'shift-jis'),
  );

  console.log(result);
};
