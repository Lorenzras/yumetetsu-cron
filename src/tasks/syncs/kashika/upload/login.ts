/* eslint-disable max-len */
import path from 'path';
import {Page} from 'puppeteer';
import {Scheduler} from 'tesseract.js';
import {TStoreSettingsItem} from '../../../../config';
import fs from 'fs';
import {logger} from '../../../../utils';


export const login = async (
  page: Page,
  scheduler: Scheduler,
  {
    email,
    pass,
  }: TStoreSettingsItem,
  workerId: number,
) => {
  if (!email || !pass) {
    throw new Error('Please provide email and password. Check .env file.');
  }

  let isSuccess = false;

  do {
    console.log('Starting login to Kashika.');
    await page.goto('https://kasika.io/login.php', {waitUntil: 'domcontentloaded'});
    let cleanText = '';

    console.log('Waiting for captcha selector.');
    const testImg = await page.waitForSelector('form img', {visible: true});

    console.log('Saving captcha.');
    const testImgPath = path.join(__dirname, '__TEST__', `${workerId}.png`);
    const dir = path.dirname(testImgPath);
    fs.existsSync(dir) || fs.mkdirSync(dir, {recursive: true});

    await page.waitForTimeout(1000); // throttle by 1 second.

    console.log('Taking captcha screenshot.');
    await testImg?.screenshot({
      path: testImgPath,

    });

    console.log('Done taking screenshot.');

    // const {data: {text, confidence, tsv}} = await worker
    //  .recognize(path.join(__dirname, 'test.png'));

    const {data: {text, confidence, tsv}} = await scheduler
      .addJob('recognize', testImgPath);


    console.log(text, 'conf: ' + confidence, 'tsv:\n' + tsv);
    cleanText = text.trim().replaceAll(' ', '');

    if (cleanText.length === 4 && confidence >= 90) {
      logger.info(`Possible match ${text}`);
      await page.$('input[type=email]').then(async (input)=>{
        await input?.click({clickCount: 3});
        await input?.type(email);
      });

      await page.$('input[type=password]').then(async (input)=>{
        await input?.click({clickCount: 3});
        await input?.type(pass);
      });

      await page.$('input[name=captcha]').then(async (input)=>{
        await input?.click({clickCount: 3});
        await input?.type(cleanText);
      });

      logger.info('Clicking submit');

      await Promise.all([
        page.waitForNavigation(),
        page.click('button[type=submit]'),
      ]);

      console.log('Done waiting for navigation.');

      isSuccess = await Promise.race([
        page.waitForSelector('div#kasika-table', {visible: true}).then(()=>true),
        page.waitForSelector('div[role=alert]', {visible: true}).then(()=>false),
      ]);
    } else {
      isSuccess = false;
    }
  } while (!isSuccess);

  // await worker.terminate();
};

// failed, 92, 84, 82, 67, 88, 56, 23, 55, 12, 6, 3, 9, 17
// success, 94, 85, 92, 81, 95, 96, 84, 95, 91, 93
