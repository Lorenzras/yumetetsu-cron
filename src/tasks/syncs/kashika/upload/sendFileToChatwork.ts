
import fs from 'fs';
import FormData from 'form-data';
import axios from 'axios';
import {logger} from '../../../../utils';

// test = 225800073
// rpachat = 213232379
// yumetetsu = 6732051

const roomId = '6732051';

export const sendFileToChatwork = async ({
  storeName,
  filePath,
  fileDetails,
  totalCount,
  cwToken = process.env.CW_TOKEN,
}:
{
  fileDetails: Record<string, number>,
  filePath: string,
  totalCount: number,
  storeName: string,
  cwToken?: string
},
) => {
  logger.info('sending to chatwork.');

  const formData = new FormData();
  const strFileDetails = Object.entries(fileDetails).map(([key, val]) => {
    console.log(key);
    return `${key}： ${val}件`;
  }).join('\n');

  const message = [
    `[info][title]${storeName} → KASIKA: スキップされる顧客データ（エラー）[/title]`,
    `${totalCount ?? 0}件の中、以下はスキップされました。\n`,
    `${strFileDetails}\n`,
    `[hr]`,
    `詳しくは添付のファイルをダウンロードしてください。`,
    `※エクセルで開くと見やすいです。[/info]`,
  ].join('\n');

  formData.append('message', message);
  formData.append('file', fs.createReadStream(filePath));

  const url = `https://api.chatwork.com/v2/rooms/${roomId}/files`;


  const headers = {
    'Accept': 'application/json',
    'X-ChatWorkToken': cwToken,
  };


  await axios({
    method: 'post',
    url: url,
    headers: headers,
    data: formData,
  });

/*   return await fetch(url, options)
    .then((res) => res.json())
    .then((json) => console.log(json))
    .catch((err) => console.error('error:' + err)); */
};

/* const fs = require('fs');
const FormData = require('form-data');
const fetch = require('node-fetch');
const formData = new FormData();

formData.append('message', 'hello');
formData.append('file', fs.createReadStream('1343.csv'));

const url = 'https://api.chatwork.com/v2/rooms/225800073/files';
const options = {
  method: 'POST',
  headers: {
    Accept: 'application/json',
    'X-ChatWorkToken': '7bc795ef967064f642aa70956cde3cad'
  }
};

options.body = formData;

fetch(url, options)
  .then(res => res.json())
  .then(json => console.log(json))
  .catch(err => console.error('error:' + err)); */
