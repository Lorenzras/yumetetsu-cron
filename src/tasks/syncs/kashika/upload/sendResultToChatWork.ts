import axios from 'axios';
import FormData from 'form-data';
import {storeSettings} from '../../../../config';

const roomId = '6732051';

export const sendResultToChatWork = async (
  record: Record<string, Record<string, string>[]>,
) => {
  const selfUnread = 1;

  const url = `https://api.chatwork.com/v2/rooms/${roomId}/messages`;

  let totalLength = 0;
  const content = Object.entries(record).map(([key, perStoreRecs]) => {
    const storeName = storeSettings[key as keyof typeof storeSettings].name;
    totalLength += perStoreRecs.length;
    return `${storeName}：${perStoreRecs.length}件`;
  }).join('\n');

  const message = [
    `[info][title]KASIKAアップロード[/title]`,
    content,
    `[hr]`,
    `合計：${totalLength}件`,
    `[/info]`,
  ].join('\n');


  // formData.append('message', message);
  const params = `body=${message}&self_unread=${selfUnread}`;
  const headers = {
    'Accept': 'application/json',
    'X-ChatWorkToken': process.env.CW_TOKEN,
  };

  await axios({
    method: 'post',
    url: url,
    headers: headers,
    data: params,
  });
};
