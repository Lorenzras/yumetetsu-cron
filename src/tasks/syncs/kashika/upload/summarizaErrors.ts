export const summarizeErrors = (rows: Record<string, string>[]) => {
  return rows.reduce((accu, prev)=>{
    const errorInfo = prev['エラー内容'].replaceAll(/\d:\s/g, '');
    if (!accu[errorInfo]) accu[errorInfo] = 0;
    accu[errorInfo] += 1;

    return accu;
  }, {} as Record<string, number>);
};
