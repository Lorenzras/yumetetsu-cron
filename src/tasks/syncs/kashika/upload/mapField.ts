/* eslint-disable max-len */
import {Page} from 'puppeteer';
import {cleanStoreName} from './cleanStoreName';

/**
 * マッピング設定
 *
 * @param param
 * @param param.page
 * @param param.csvField The csvField
 * @param param.kasikaField Array of strings that will match option. Used for retrieving value
 *
 */
export const mapField = async ({
  page, csvField, store,
} :
{
  page: Page,
  csvField: '顧客種別' | '顧客ステータス',
  store: string
}) => {
  await page.waitForSelector(`input[value="${csvField}"]`);


  const cleanStore = cleanStoreName(store);

  await page.evaluate((
    csvField: string,
    cleanStore: string,
  )=>{
    const selectEl = $(`input[value="${csvField}"]`)
      .closest('td')
      .find('.jsSelectRow select:first');

    const optValue = selectEl
      .find('option')
      .filter((_, el) => {
        return [cleanStore, csvField]
          .every((str) => $(el).text().includes(str));
      })
      .attr('value');

    if (!optValue) throw new Error(`Failed to find option ${[csvField, cleanStore].join(',')}`);
    selectEl.val(optValue);
  }, csvField, cleanStore);
};
