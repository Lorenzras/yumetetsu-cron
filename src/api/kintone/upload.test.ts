import {APP_IDS} from '.';
import {uploadAll} from './upload';
import {faker} from '@faker-js/faker';

const data: Partial<CustomersType>[] = [
  {
    '顧客番号': {value: faker.random.alphaNumeric()},
    '氏名': {value: faker.name.findName()},
    '氏名_フリガナ': {value: 'test only'},
  },
];


describe('All Records', ()=>{
  it('are succesfully uploaded.', async ()=>{
    const result = await uploadAll(APP_IDS['customers'], data )
      .catch((err) => console.log(err));
    console.log(result);
    expect(result).toMatchSnapshot();
  });
});
