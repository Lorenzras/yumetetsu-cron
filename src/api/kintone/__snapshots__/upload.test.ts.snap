// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All Records are succesfully uploaded. 1`] = `
Object {
  "records": Array [
    Object {
      "id": "39433",
      "revision": "1",
    },
  ],
}
`;
